cocotb/share/lib/embed/embed.cpp

```diff
@@ -0,0 +1,133 @@
+// Copyright cocotb contributors
+// Licensed under the Revised BSD License, see LICENSE for details.
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <embed.h>
+#include <cocotb_utils.h>   // xstr, utils_dyn_open, utils_dyn_sym
+#include <gpi.h>            // gpi_event_t
+#include <cstdlib>          // getenv
+#if ! defined(__linux__) && ! defined(__APPLE__)
+#include <windows.h>        // Win32 API for loading the embed impl library
+#include <string>           // string
+#endif
+
+#ifndef EMBED_IMPL_LIB
+#error "Name of embed implementation library required"
+#else
+#define EMBED_IMPL_LIB_STR xstr(EMBED_IMPL_LIB)
+#endif
+
+
+static void (*_embed_init_python)(void);
+static void (*_embed_sim_cleanup)(void);
+static int (*_embed_sim_init)(int argc, char const * const * argv);
+static void (*_embed_sim_event)(gpi_event_t level, const char *msg);
+
+static bool init_failed = false;
+
+
+#define PRINT_ERROR_WINDOZE(log_fmt) {\
+    LPSTR msg_ptr;\
+    if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_ALLOCATE_BUFFER, NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_SYS_DEFAULT), reinterpret_cast<LPSTR>(&msg_ptr), 255, NULL)) {\
+        LOG_ERROR(log_fmt ": %s", msg_ptr);\
+        LocalFree(msg_ptr);\
+    } else {\
+        LOG_ERROR(log_fmt);\
+    }}
+
+
+static void * load_embed_impl_lib()
+{
+#if ! defined(__linux__) && ! defined(__APPLE__)
+    HMODULE hm = NULL;
+
+    // get a handle to the module
+    if (!GetModuleHandleExA(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, reinterpret_cast<LPCSTR>(&embed_init_python), &hm))
+    {
+        PRINT_ERROR_WINDOZE("Unable to open embed library");
+    }
+
+    // get the path to the module
+    char path[MAX_PATH];
+    if (!GetModuleFileNameA(hm, path, sizeof(path)))
+    {
+        PRINT_ERROR_WINDOZE("Unable to get filepath of embed library");
+    }
+
+    auto embed_lib_filepath = std::string(path);
+
+    // strip filename from path
+    auto embed_lib_dirpath = embed_lib_filepath.substr(0, embed_lib_filepath.rfind("\\"));
+
+    // combine the directory from the last step and the supplied filename
+    auto embed_impl_lib_path = embed_lib_dirpath + "\\" + EMBED_IMPL_LIB_STR;
+
+    return utils_dyn_open(embed_impl_lib_path.c_str());
```

Comment:
Implementation gleaned from [this answer](https://stackoverflow.com/a/65165593/6614127) and [this answer](https://stackoverflow.com/a/6924332/6614127).

While the use of `PathRemoveFileSpec` and `PathCombine` are recommended, it would require we link against `shwapi` library. We can expect the path returned from `GetModuleFileNameA` to be canonicalized, so spliting on `\` and appending `\libcocotb.dll` should "just work". We can combat that in the future if it becomes a problem.